google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionPool.java

```diff
@@ -838,27 +888,29 @@ private void handlePrepareSessionFailure(SpannerException e, PooledSession sessi
       Waiter waiter = readWaiters.poll();
       while (waiter != null) {
         waiter.put(
-            SpannerExceptionFactory.newSpannerException(
+            newSpannerException(
                 ErrorCode.INTERNAL, "Client has been closed"));
         waiter = readWaiters.poll();
       }
       waiter = readWriteWaiters.poll();
       while (waiter != null) {
         waiter.put(
-            SpannerExceptionFactory.newSpannerException(
+            newSpannerException(
                 ErrorCode.INTERNAL, "Client has been closed"));
         waiter = readWriteWaiters.poll();
       }
       closureFuture = SettableFuture.create();
       retFuture = closureFuture;
-      pendingClosure = totalSessions + numSessionsBeingCreated;
-      if (pendingClosure == 0) {
-        closureFuture.set(null);
-      } else {
-        for (final PooledSession session :
-            Iterables.consumingIterable(Iterables.concat(readSessions, writePreparedSessions))) {
-          closeSessionAsync(session.delegate);
+      pendingClosure =
+          totalSessions() + numSessionsBeingCreated + 1 /* For pool maintenance thread */;
+      poolMaintainer.close();
+      readSessions.clear();
+      writePreparedSessions.clear();
+      for (final PooledSession session : ImmutableList.copyOf(allSessions.values())) {
+        if (session.leakedException != null) {
+          logger.log(Level.WARNING, "Leaked session", session.leakedException);
```

Comment:
Do you think it would be useful to expose some way of dumping this on demand? For example, if someone wanted to see what sessions were currently in use without shutting down the client entirely.